sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+of the required work-group or sub-group size).
+
+As we will see later, it will be very helpful to decorate all APIs in DPC++
+headers that correspond to optional kernel features with the
+`[[sycl::requires()]]` attribute.  For example, the declaration of the
+`sycl::half` type would look like this:
+
+```
+using half [[sycl::requires(has(aspect::fp16))]] = cl::sycl::detail::half_impl::half;
+```
+
+In cases where the optional feature corresponds to use of a class (e.g.
+`sycl::atomic_ref`), the declaration can look like this:
+
+```
+template</*...*/>
+class [[sycl::requires(has(aspect::atomic64))]] atomic_ref {
+  /* ... */
+};
+```
+
+(In reality, we can use partial specialization tricks to decorate `atomic_ref`
+with the attribute only when the underlying type is 64-bits.  See ["Appendix:
+Adding an attribute to 8-byte `atomic_ref`"][4].)
+
+[4]: <#appendix-adding-an-attribute-to-8-byte-atomic_ref>
+
+In cases where the optional feature corresponds to a function, we can decorate
+the function's declaration with the attribute like so (demonstrating a
+hypothetical AMX multiplication extension):
+
+```
+[[sycl::requires(has(aspect::ext_intel_amx))]]
+void amx_multiply();
+```
+
+These attributes provide an explicit mapping between each optional kernel
+feature and its associated aspect.
+
+Unfortunately, the fundamental type `double` is also an optional kernel
+feature.  Since there is no type alias for `double`, there is no convenient
+place to add an attribute.  Instead, the front-end device compiler must behave
+as though there was an implicit `[[sycl::requires(has(aspect::fp64))]]`
+attribute for any device code that uses the `double` type.
+
+Note that the usage of `[[sycl::requires()]]` is slightly expanded here beyond
+the specified usage in the SYCL 2020 specification because we allow the
+attribute also on type alias declarations and class declarations.  If a device
+function does any of the following with a type alias or class that was so
+decorated, the function is assumed to "use the aspects" listed in the
+attribute:
+
+* Declares a variable of that type.
+* Has a formal parameter declared with that type.
+* Returns that type.
+
+This also includes any qualified version of the type.
+
+**TODO**: This language is not very precise.  The intent is to include most
+uses of the type, except for cases like `sizeof(T)` or `decltype(T)`.  Help
+appreciated on tightening the wording here.
+
+**TODO**: Would it be better to use a different attribute name when decorating
+types, rather than expanding the meaning of `[[sycl::requires()]]`?  If we did
+this, the new attribute would become an internal DPC++ implementation detail;
+we would not add it to the SYCL specification.
+
+### Implementing diagnostics in the DPC++ front-end
+
+As noted above, the front-end device compiler must behave as though there is an
+implicit `[[sycl::requires(has(aspect::fp64))]]` attribute on any use of the
+`double` type in device code.
```

Comment:
Presumably there's something we should do about `long double` here?

Reply:
Actually, `long double` is not an **optional** device feature.  The SYCL spec lists all of the [allowed fundamental types](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:scalartypes) in device code, and this does not include `long double`.  Thus, `long double` is one of the C++ features that is disallowed in device code.  Ideally, some part of the DPC++ compiler would emit an error message whenever device code uses one of these disallowed features.  (Maybe we do this already?)  However, that is not the topic of this design document.

Reply:
We should disallow long double.  as @gmlueck says, not an allowed fundamental type.  As an aside, we would even like to discourage the use of that type in host code, and have people use __float128 instead.

Reply:
Oh, it turns out we do disallow it... sort of. :-(
```
template <typename Ty, typename Uy>
__attribute__((sycl_kernel)) void foo(long double this_one_is_fine) { // ok
  long double but_this_one_is_bad; // error
}
```
So it looks like there's some work left to be done on that checking. Regardless, agreed that's not the point of the design doc, so we can Resolve this comment. Thanks for the explanation!